### PR TITLE
Improvements/transaction docs large refactorings

### PR DESCRIPTION
**Pull request by astubbs**
- [ ] Changelog

Blocked by:
- #398

### CODE REVIEW

**Review by astubbs — COMMENTED** [review]
LGTM

File: parallel-consumer-core/src/test/java/io/confluent/csid/utils/TrimListRepresentationTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.stream.IntStream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+/**
+ * @author Antony Stubbs
+ * @see TrimListRepresentation
+ */
+@Tag("transactions")
```
Comment (rkolesnev) [inline parallel-consumer-core/src/test/java/io/confluent/csid/utils/TrimListRepresentationTest.java:23]: why is it tagged as Transactions? i can see its moved from Transaction related test class but they seem to be completely isolated utility tests?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java [hunk 2]
```diff
@@ -58,40 +58,43 @@ public void poll(Consumer<PollContext<K, V>> usersVoidConsumptionFunction) {
...
+                context -> userFunctionWrap(userFunction, context);
 
-            if (recordListToProduce.isEmpty()) {
-                log.debug("No result returned from function to send.");
-            }
-            log.trace("asyncPoll and Stream - Consumed a record ({}), and returning a derivative result record to be produced: {}", context, recordListToProduce);
+        supervisorLoop(wrappedUserFunc, callback);
+    }
+
+    private List<ConsumeProduceResult<K, V, K, V>> userFunctionWrap(Function<PollContext<K, V>, List<ProducerRecord<K, V>>> userFunction,
```
Comment (rkolesnev) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:73]: nitpick - better name for the function? producingWrapper? something to tie in with the fact that it is the part that does produce after userFunction processed consumed messages? 
as it stands it is not really telling me anything about what that function does.

Comment (rkolesnev) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:73]: May be something along the lines of 
```
Function<PollContextInternal<K, V>, List<ConsumeProduceResult<K, V, K, V>>> producingUserFunctionWrapper = context -> {
    List<ProducerRecord<K, V>> recordListToProduce = carefullyRun(userFunction, context.getPollContext());
    return produceMessages(recordListToProduce, context));
    };
```
or just renaming maybe
```
Function<PollContextInternal<K, V>, List<ConsumeProduceResult<K, V, K, V>>> producingUserFunctionWrapper = context -> processAndProduceResults(userFunction, context);
```

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import java.lang.reflect.Method;
+import java.util.Map;
+
+/**
+ * Our extension of the standard Consumer to mostly add some introspection functions and state tracking.
+ *
+ * @author Antony Stubbs
+ */
+@RequiredArgsConstructor
+public class ProducerWrap<K, V> implements Producer<K, V> {
```
Comment (rkolesnev) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerWrap.java:30]: Maybe ProducerWrapper ? or WrappedProducer?

**Review by rkolesnev — APPROVED** [review]
LGTM - couple of naming / flow comments above.

File: parallel-consumer-core/src/test/java/io/confluent/csid/utils/TrimListRepresentationTest.java (hunk 1 above)
Comment (astubbs) [reply parallel-consumer-core/src/test/java/io/confluent/csid/utils/TrimListRepresentationTest.java:23]: PR was created after, the refactor is apart of the transactions work. but you're right - I'll remove it.

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java (hunk 2 above)
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:73]: slack discussion

**Pull request merged**